Feature/Popover Definitions

Bouncey opened the pull request:
<!-- FreeCodeCamp Pull Request Template -->

<!-- IMPORTANT Please review https://github.com/FreeCodeCamp/FreeCodeCamp/blob/staging/CONTRIBUTING.md for detailed contributing guidelines -->

<!-- Help with PRs can be found at https://gitter.im/FreeCodeCamp/HelpContributors -->

<!-- Make sure that your PR is not a duplicate -->
#### Pre-Submission Checklist

<!-- Go over all points below, and put an `x` in all the boxes that apply. -->

<!-- All points should be checked, otherwise, read the CONTRIBUTING guidelines from above-->

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](https://github.com/FreeCodeCamp/FreeCodeCamp/wiki/git-rebase#squashing-multiple-commits-into-one) them into one commit).
- [x] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

<!-- What type of change does your code introduce? Put an `x` in the box that applies. -->
- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

<!-- Go over all points below, and put an `x` in all the boxes that apply. -->

<!-- If you're unsure about any of these, don't hesitate to ask in the Help Contributors room linked above. We're here to help! -->
- [x] Tested changes locally.
- [x] Closes currently open issue (replace XXXX with an issue no): Closes #4117
#### Description

<!-- Describe your changes in detail -->

Added popover definitions to one challenge only (`Say Hello to HTML Elements`) for review/feedback.

Bouncey commented:
One concern I have is doing it this way makes the code look messy and harder to understand. I have a fair idea how to smarten it up with angular (custom directive), but not react.

Bouncey commented:
Reworking, smartening up the look of the popover and the code needed. 

Please block until amended @erictleung @raisedadead

erictleung commented:
@Bouncey thanks for the update. Some feedback on your changes, your solution so far (not sure if you were plan on changing it) doesn't appear to be scalable because the definition is within the challenge itself. What was suggested in that issue discussion was having a separate JSON file with all the definitions. Let us know if you have any questions.

Bouncey commented:
Just updated with a different version. I removed most of what I could from the challenge seed, but the popover needs to come from either a `<button>` tag or `<a>` tag.

I aim to split the specific CSS into a different file as well for the final commit.

`def.js` was the easiest way I could see of populating the popover with the required content, and still have some readability in the challenge seed.

Bouncey commented:
Please QA

bayuanderas commented:
http://whow.simdif.com
Pada tanggal 10 Jun 2016 20.01, "CamperBot" <notifications@github.com>
menulis:

> @Bouncey <https://github.com/Bouncey> updated the pull request.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/FreeCodeCamp/FreeCodeCamp/pull/8917#issuecomment-225175536>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AHd1apQxwKKcZuHGJIvLEI96GNE7J5oAks5qKWAOgaJpZM4ItBTX>
> .
>

erictleung commented:
@Bouncey there appears to be some merge conflicts now. I would rebase this branch onto an updated repo. If you've gotten your remotes set up correctly, this code should help you.

```bash
git checkout staging
git pull --rebase upstream staging
git checkout Feature/Popover-Def-HTML-CSS
git rebase staging
```

From there, you should be able to find the conflicts and make changes accordingly.

cc/ @FreeCodeCamp/issue-moderators any feedback on how @Bouncey is implementing the definitions?

Bouncey commented:
@erictleung @raisedadead 

I have 
`git commit --amend`
`git push -f`

but TravisCI hasn't fired?

Bouncey commented:
OK, please QA

atjonathan commented:
@FreeCodeCamp/issue-moderators ^

I love the way it looks, you have a few spacing issue you could fix.
Maybe rename the file to `definitions-showHTML.js`

Commit:
Feature/Popover Definitions

raisedadead commented:
@atjonathan Looks like this needs re-basing.

BerkeleyTrue commented on client/less/main.less:
```diff
@@ -1161,6 +1161,41 @@ code {
 }
 
+// for bootstrap popover
+.pop-explain code{
```
These new css class definitions should be in their own file

Bouncey replied:
:+1:  Done

BerkeleyTrue commented on public/js/definitions-html5-css.js:
```diff
@@ -0,0 +1,91 @@
+$(document).ready(function(){
```
This file should be in client/commonFramework folder and follow the init pattern established there

Bouncey replied:
I think I'm starting to get it, would I need to start the file with something like

```
window.common = (function(global) {
  const {
    $,
    Rx: { Observable },
    common = { init: [] },
    Mousetrap
  } = global;
```
?

Bouncey replied:
:+1: Got it

BerkeleyTrue commented on server/views/challenges/showHTML.jade:
```diff
@@ -73,6 +73,9 @@ block content
                                 a#next-challenge.btn.btn-lg.btn-primary.btn-block(href="/challenges/next-challenge?id="+id) Go to my next challenge (ctrl + enter)
     include ../partials/challenge-modals
+    
+    script(src='/js/definitions-html5-css.js')
```
This should be combined into the common framework and not a separate file

Bouncey replied:
:+1:

Bouncey commented:
Just re-basing, and will amend.

Bouncey commented on seed/challenges/01-front-end-development-certification/basic-javascript.json:
```diff
@@ -3385,7 +3385,7 @@
       "title": "Golf Code",
       "description": [
-        "In the game of <a href=\"https://en.wikipedia.org/wiki/Golf\" target=\"_blank\">golf</a> each hole has a <code>par</code> for the average number of <code>strokes</code> needed to sink the ball. Depending on how far above or below <code>par</code> your <code>strokes</code> are, there is a different nickname.",
+        "In the game of <a href=\"https://en.wikipedia.org/wiki/Golf\" target=\"_blank\">golf</a> each hole has a <code>par</code> meaning the average number of <code>strokes</code> a golfer is expected to make in order to sink the ball in a hole to complete the play. Depending on how far above or below <code>par</code> your <code>strokes</code> are, there is a different nickname.",
```
I have tried re-basing and reset hard, but cannot remove this file from my commit

Bouncey commented:
Oh, I now have 24 commits..... I did something wrong...

BerkeleyTrue commented on gulpfile.js:
```diff
@@ -132,7 +132,8 @@ var paths = {
     'step-challenge',
-    'end'
+    'end',
+    'definitions-html5-css'
```
This should come before 'end'

BerkeleyTrue commented:
Looks like you brought in a bunch of commits. You may want to start over with a fresh branch

Bouncey commented:
I fixed it now

Bouncey commented:
Ok, Ill amend the gulpfile

atjonathan commented on client/commonFramework/definitions-html5-css.js:
```diff
@@ -0,0 +1,125 @@
+common = { init: [] }
+} = global;
+
+// Say Hello to HMTL Elements
```
~HMTL Elements~~ HTML Elements

atjonathan commented on client/less/popover.less:
```diff
@@ -0,0 +1,34 @@
+    font-size: inherit;
+    background-color: @brand-primary;
+    border-radius: 5px 5px 0 0;
+    color: #fff;
```
You are using tabs here, when you were using spaces earlier.

atjonathan commented on seed/challenges/01-front-end-development-certification/html5-and-css.json:
```diff
@@ -1415,7 +1416,7 @@
       "tests": [
         "assert((/cat photos/gi).test($(\"a\").text()), 'message: Your <code>a</code> element should have the <code>anchor text</code> of \"cat photos\".');",
-        "assert(/http:\\/\\/freecatphotoapp\\.com/gi.test($(\"a\").attr(\"href\")), 'message: You need an <code>a</code> element that links to <code>http&#58;//freecatphotoapp<wbr>.com</code>');",
+        "assert(/http:\\/\\/freecatphotoapp\\.com/gi.test($(\"a\").attr(\"href\")), 'message: You need an <code>a</code> element that links to <code>http&#58;//freecatphotoapp.com</code>');",
```
@Bouncey, why was `<wbr>` removed? This was recently implemented, and shouldn't be removed.

Bouncey replied:
@atjonathan I didn't? I've only just rebased the branch too?

I'll add it back in

atjonathan replied:
@Bouncey, probably due to your `git` issues earlier.

atjonathan commented on seed/challenges/01-front-end-development-certification/html5-and-css.json:
```diff
@@ -4532,7 +4533,7 @@
       ],
       "tests": [
-        "assert($(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\", 'message: Your <code>body</code> element should have a black background.');",
+        "assert($(\"body\").css(\"background-color\") === \"rgb(0, 0, 0)\", 'message: Your <code>body</code> element should have an black background.');",
```
~~an black background.~~ **a** black background.

Bouncey replied:
I think this is related to the <wbr> issue as well

atjonathan replied:
Yeh it is :smile:

atjonathan commented:
@Bouncey, these look good! You could start working on the next bunch of challenges :+1:

Bouncey commented:
I have read through every challenge in `HTML5 and CSS`, `Bootstrap` and `jquery'. I'm stuggling to find new concepts that aren't well explained in the challenge instructions.

atjonathan commented:
@Bouncey, it's possible we could eliminate some of the instructions, and replace it with popovers.

Bouncey commented:
@atjonathan You'll lose some of the context of the instructions, so you end up having to re-write the challenge in order to have a popover. A lot of effort...

Bouncey commented:
I think the format is established enough for other new concepts to be caught with issues

atjonathan commented:
@Bouncey, can this be merged or are you going to add more?

BerkeleyTrue commented:
I'd like to take a look at this before it's merged

Bouncey commented:
Any feedback @BerkeleyTrue ?

BerkeleyTrue commented:
Sorry about that @Bouncey I haven't had the chance. My top priority is moving everything over to React.

Do you mind adding some gifs of what it looks like?

@atjonathan Did you QA this?

Bouncey commented:
@BerkeleyTrue I understand you not getting round to it, not a problem.

![anim](https://cloud.githubusercontent.com/assets/18572518/16535271/ff9f4976-3fdd-11e6-9a46-f131a59e39aa.gif)

Bouncey commented:
@BerkeleyTrue And mobile

![anim](https://cloud.githubusercontent.com/assets/18572518/16535402/e135222a-3fde-11e6-9060-708e5e01d3f3.gif)

atjonathan commented:
I will `QA` this later today @BerkeleyTrue.

atjonathan commented:
### **LGTM :+1:**
Great job on this @Bouncey :tada: :stuck_out_tongue_winking_eye: 
@QuincyLarson, please take a look before this is merged...

atjonathan commented:
Alright, merging as looks good and the user interface can always be changed later :+1:

BerkeleyTrue commented:
@atjonathan This was waiting on @QuincyLarson to comment

BerkeleyTrue commented:
I've reverted the changes. @Bouncey please feel free to reopen the PR and wait for @QuincyLarson's comment.

Happy Coding

atjonathan commented:
@BerkeleyTrue, Np I was merging after a while of inactivity and multiple issue mods had tested.
